karabo_data/geometry2.py

```diff
@@ -205,6 +208,83 @@ def inspect(self, frontview=True):
 
         return ax
 
+    @classmethod
+    def from_crystfel_geom(cls, filename):
+        """Read a CrystFEL format (.geom) geometry file.
+
+        Returns a new geometry object.
+        """
+        geom_dict = load_crystfel_geometry(filename)
+        modules = []
+        for p in range(cls.n_modules):
+            tiles = []
+            modules.append(tiles)
+            for a in range(cls.n_asics_per_module):
+                d = geom_dict['panels']['p{}a{}'.format(p, a)]
+                tiles.append(GeometryFragment.from_panel_dict(d))
+        return cls(modules, filename=filename)
+
+    def _get_rigid_groups(self, nquads=4):
+        """Create ridget stings for rigid groups definiton."""
+
+        quads = ','.join(['q{}'.format(q) for q in range(nquads)])
+        modules = ','.join(['p{}'.format(p) for p in range(self.n_modules)])
+
+        prod = product(range(self.n_modules), range(self.n_asics_per_module))
+        rigid_group = ['p{}a{}'.format(p, a) for (p, a) in prod]
+        rigid_string = '\n'
+
+        for nn, rigid_group_q in enumerate(np.array_split(rigid_group, nquads)):
+            rigid_string += 'rigid_group_q{} = {}\n'.format(nn, ','.join(rigid_group_q))
+        rigid_string += '\n'
+        for nn, rigid_group_p in enumerate(np.array_split(rigid_group, self.n_modules)):
+            rigid_string += 'rigid_group_p{} = {}\n'.format(nn, ','.join(rigid_group_p))
+
+        rigid_string += '\n'
+
+        rigid_string += 'rigid_group_collection_quadrants = {}\n'.format(quads)
+        rigid_string += 'rigid_group_collection_asics = {}\n\n'.format(modules)
+        return rigid_string
+
+    def write_crystfel_geom(self, filename, adu_per_ev=0.0075, clen=0.119,
+                            photon_energy=10235):
```

Comment:
I'd rather not have default values here. Presumably they will rarely be right, especially across different detectors and instruments, but it's very easy for someone who doesn't know the real values to assume that the defaults are probably a good starting point. They might either waste time trying to do something that can't work, or even produce invalid results because they've got bad parameters.

Of course it's ultimately the scientist's responsibility to notice things like this and find the correct values. But our software tools should make it easy to do the right thing and hard to do the wrong thing.

I don't think the karabo_data API should require these parameters - just have `None` for the defaults and write them into the geometry file only if they're specified.

Reply:
I'd also make them keyword-only arguments, which works like this:

https://github.com/European-XFEL/karabo_data/blob/9728c5b9be6fbbec6ccbc4ce09d99421e6ba56ba/karabo_data/geometry2.py#L816-L817

Reply:
Agreed!